ouroboros-network-api/src/Ouroboros/Network/PeerSelection/LedgerPeers/Type.hs

```diff
@@ -15,17 +22,134 @@ module Ouroboros.Network.PeerSelection.LedgerPeers.Type
+  deriving (Eq, Show)
+
+-- |Public API to access snapshot data. Currently access to only most recent version is available.
+-- Nonetheless, serialisation from the node into JSON is supported for older versions via internal
+-- api so that newer CLI can still support older node formats.
+--
+pattern LedgerPeerSnapshot :: (WithOrigin SlotNo, [(AccPoolStake, (PoolStake, NonEmpty RelayAccessPoint))])
+                           -> LedgerPeerSnapshot
+pattern LedgerPeerSnapshot payload <- LedgerPeerSnapshotV1 payload where
+  LedgerPeerSnapshot payload = LedgerPeerSnapshotV1 payload
+
+{-# COMPLETE LedgerPeerSnapshot #-}
+
+-- | In case the format changes in the future, this function provides a migration functionality
+-- when possible.
+--
+migrateLedgerPeerSnapshot :: LedgerPeerSnapshot
+                          -> Maybe (WithOrigin SlotNo, [(AccPoolStake, (PoolStake, NonEmpty RelayAccessPoint))])
+migrateLedgerPeerSnapshot (LedgerPeerSnapshotV1 lps) = Just lps
+
+instance ToJSON LedgerPeerSnapshot where
+  toJSON (LedgerPeerSnapshotV1 (slot, pools)) =
+    object [ "Version" .= (1 :: Int)
+           , "SlotNo" .= slot
+           , "BigLedgerPools" .= [ object [ "AccStake" .= show accStake
+                                          , "ReStake"  .= show stake
+                                          , "Relays"   .= relays]
+                                 | (AccPoolStake accStake, (PoolStake stake, relays)) <- pools]]
+
+instance FromJSON LedgerPeerSnapshot where
+  parseJSON = withObject "LedgerPeerSnapshot" $ \v -> do
+    vNum <- v .: "Version"
+    parsedSnapshot <-
+      case (vNum :: Int) of
+        1 -> do
+          slot <- v .: "SlotNo"
+          bigPools <- v .: "BigLedgerPools"
+          bigPools' <- (forM bigPools . withObject "BigLedgerPools" $ \poolV -> do
+            accStake <-
+              poolV .: "AccStake"
+              >>= (\case
+                    Just accStake -> pure accStake
+                    Nothing    ->
+                      fail "Invalid AccStake value")
+                  . readMaybe
+            reStake <-
+              poolV .: "ReStake"
+              >>= (\case
+                    Just reStake -> pure reStake
+                    Nothing    ->
+                      fail "Invalid ReStake value")
+                  . readMaybe
+            relays <- poolV .: "Relays"
+            return (accStake, (reStake, relays))) <?> Key "BigLedgerPools"
+
+          return $ LedgerPeerSnapshotV1 (slot, bigPools')
+        _ -> fail $ "Network.LedgerPeers.Type: parseJSON: failed to parse unsupported version " <> show vNum
+    case migrateLedgerPeerSnapshot parsedSnapshot of
+      Just payload -> return $ LedgerPeerSnapshot payload
+      Nothing      -> fail "Network.LedgerPeers.Type: parseJSON: failed to migrate big ledger peer snapshot"
+
+-- | Hand cranked CBOR instances to facilitate CDDL spec
+--
+instance {-# OVERLAPPING #-} ToCBOR (WithOrigin SlotNo) where
```

Comment:
Why do you need OVERLAPPING pragma?

Reply:
There is a ```ToCBOR (WithOrigin a)``` instance in cardano-slotting - I've made a patch PR to make it OVERLAPPABLE there. I'll hold off on merging this one until it gets approved. If not, I guess I can just create a newtype wrapper.

Reply:
If you already have an instance for `WithOrigin` than it is not our responsibility to provide one since that's done in a different place. We should provide instance for types we own, so in this case we can't do much the way I see it! Sorry if I was misleading regarding CBOR instances, but I don't think we should go through the trouble of making these overlappable. They provide the `WithOrigin` instance and we just either need to provide the `SlotNo` instance or get it from somewhere if we don't own that type

Reply:
Their definition looks like this:
`instance {-# OVERLAPPABLE #-} (Serialise t, Typeable t) => ToCBOR (WithOrigin t) where
  toCBOR = encode`
I guess to have full control of the encoding this is not really useable. I've created a newtype wrapper instead.